File: sparql-scala/src/main/scala/org/aksw/commons/sparql/core/SparqlTemplate.scala
package org.aksw.commons.sparql.core

import org.apache.velocity.app.VelocityEngine
import java.io.StringWriter
import org.apache.velocity.runtime.RuntimeConstants
import org.apache.velocity.runtime.resource.loader.ClasspathResourceLoader
import org.apache.velocity.{Template, VelocityContext}

import java.util.ArrayList

/**
 * User: Sebastian Hellmann - http://bis.informatik.uni-leipzig.de/SebastianHellmann 
 * Date: 04.04.11
 */

class SparqlTemplate(val template: Template) {

  def this(classpathFile: String) = this (SparqlTemplate.ve.getTemplate(classpathFile))

  var limit = 1000
  var from = new java.util.HashSet[String]()
  var filter = new java.util.HashSet[String]()
  val velocityContext = new VelocityContext
  var usePrefixes = false

  def setLimit(i: Int) {
    limit = i
  }

  def setUsePrefixes(b:Boolean) = (usePrefixes =b)

  def addFrom(s: String) {
    from.add(s)
  }

  def addFrom(s: java.util.Collection[String]) {
    from.addAll(s)
  }

  def addFilter(s: String) {
    filter.add(s)
  }

  def addFilter(s: java.util.Collection[String]) {
    filter.addAll(s)
  }

  def getVelocityContext : VelocityContext = (velocityContext)

  def getQuery(): String = {

    velocityContext.put("limit", limit);

    if (!from.isEmpty) {
      velocityContext.put("fromList", new ArrayList[String](from));
    }
    if (!filter.isEmpty) {
      velocityContext.put("filterList", new ArrayList[String](filter))
    }

    if(usePrefixes){
      velocityContext.put("prefix", usePrefixes)
    }

    val writer = new StringWriter
    template.merge(velocityContext, writer)
    writer.toString
  }

}

object SparqlTemplate extends Application {

  lazy val ve: VelocityEngine = {
    val tmp = new VelocityEngine
    tmp.setProperty(RuntimeConstants.RESOURCE_LOADER, "classpath")
    tmp.setProperty("classpath.resource.loader.class", classOf[ClasspathResourceLoader].getName)
    //TODO register these as user directives
    //tmp.setProperty("userdirective", org.apache.velocity.tools.generic.directive.Ifnull)
    //tmp.setProperty("userdirective", org.apache.velocity.tools.generic.directive.Ifnotnull)
    tmp.init
    tmp
  }

  /*lazy val map = Map[String, Template]("org.aksw.commons.sparqltemplates/allClasses.vm" -> allClasses)

  lazy val allClasses = ve.getTemplate("org.aksw.commons.sparqltemplates/allClasses.vm")

  lazy val classesOfInstance = ve.getTemplate("org.aksw.commons.sparqltemplates/classesOfInstance.vm")
  lazy val instancesOfClass = ve.getTemplate("org.aksw.commons.sparqltemplates/instancesOfClass.vm")


  def allClasses(context: VelocityContext): String = (doit(allClasses, context))

  def classesOfInstance(context: VelocityContext): String = (doit(classesOfInstance, context))

  def instancesOfClass(context: VelocityContext): String = (doit(instancesOfClass, context))


  def doit(t: Template, context: VelocityContext): String = {
    val writer = new StringWriter
    t.merge(context, writer)
    writer.toString
  }

  override def main(args: Array[String]): Unit = {

    val context = new VelocityContext
    context.put("name", "world")

    println(classesOfInstance(context))
    //getClasses(new VelocityContext())

  }
         */

}

File: sparql/src/main/java/org/aksw/commons/sparql/core/SparqlTemplate.java
package org.aksw.commons.sparql.core;


import org.apache.velocity.Template;
import org.apache.velocity.VelocityContext;
import org.apache.velocity.app.VelocityEngine;
import org.apache.velocity.runtime.RuntimeConstants;
import org.apache.velocity.runtime.resource.loader.ClasspathResourceLoader;
import org.slf4j.LoggerFactory;

import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

/**
 * User: Sebastian Hellmann - http://bis.informatik.uni-leipzig.de/SebastianHellmann
 * Date: 04.04.11
 */

public class SparqlTemplate {
    private static final org.slf4j.Logger logger = LoggerFactory.getLogger(SparqlTemplate.class);

    private static VelocityEngine ve = null;

    public static VelocityEngine getClassPathEngine() {
        if (ve == null) {
            try {
                ve = new VelocityEngine();
                ve.setProperty(RuntimeConstants.RESOURCE_LOADER, "classpath");
                ve.setProperty("classpath.resource.loader.class", ClasspathResourceLoader.class.getName());
                ve.init();
            } catch (Exception e) {
                logger.error("", e);
            }
        }
        return ve;
    }

    public static SparqlTemplate getInstance(String classpathFile) throws Exception {
        return new SparqlTemplate(getClassPathEngine().getTemplate(classpathFile));
    }


    private Template template;
    private VelocityContext velocityContext = new VelocityContext();
    //TODO this needs to be implemented better
    private boolean usePrefixes = true;
    private int limit = 1000;
    private Set<String> from = new HashSet<String>();
    private Set<String> filter = new HashSet<String>();

    public SparqlTemplate(Template template) {
        this.template = template;
    }


    public void addFrom(String s) {
        from.add(s);
    }

    public void addFrom(Collection<String> c) {
        from.addAll(c);
    }

    public void addFilter(String s) {
        filter.add(s);
    }

    public void addFilter(Collection<String> c) {
        filter.addAll(c);
    }


    public VelocityContext getVelocityContext() {
        return velocityContext;
    }


    public boolean isUsePrefixes() {
        return usePrefixes;
    }

    public void setUsePrefixes(boolean usePrefixes) {
        this.usePrefixes = usePrefixes;
    }

    public int getLimit() {
        return limit;
    }

    public void setLimit(int limit) {
        this.limit = limit;
    }

    public String getQuery() {
        velocityContext.put("limit", limit);

        if (!from.isEmpty()) {
            velocityContext.put("fromList", new ArrayList<String>(from));
        }
        if (!filter.isEmpty()) {
            velocityContext.put("filterList", new ArrayList<String>(filter));
        }

        if (usePrefixes) {
            velocityContext.put("prefix", usePrefixes);
        }

        StringWriter writer = new StringWriter();
        try {
            template.merge(velocityContext, writer);
        } catch (Exception e) {
            logger.error("", e);
        }
        return writer.toString();
    }

}

